### PR TITLE
Enable iOS/tvOS simulators on arm64

### PR DESCRIPTION
**Pull request by steveisok**

### CODE REVIEW

**Comment by steveisok** [comment]
@akoeplinger Interesting error on arm64 simulator:

https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-51536-merge-29f7a57527f24b9d93/HelloiOS/console.f22e094b.log?sv=2019-07-07&se=2021-05-10T01%3A13%3A26Z&sr=c&sp=rl&sig=y793uEF2pkU6jpQgqCMvUmvrDpUIYbpG97c2z%2Fd7Q8A%3D

```
error MT0117: Can't launch a 32-bit app on a simulator that only supports 64-bit apps (iPhone X (iOS 14.4) - created by XHarness).
[18:13:57] dbug: 18:13:57.8468270
[18:13:57] dbug: 18:13:57.8641960   at Xamarin.Launcher.SimController.Ensure64bitApp (Xamarin.Hosting.Bindings.SimDevice device, Xamarin.Hosting.AppBundle app_bundle) [0x00060] in <fd345040cf8646afbcdd85e8867162a4>:0
[18:13:57] dbug: 18:13:57.8642500   at Xamarin.Launcher.SimController.SelectAppArchitecture (Xamarin.Hosting.Bindings.SimDevice device, Xamarin.Hosting.AppBundle app_bundle) [0x000bb] in <fd345040cf8646afbcdd85e8867162a4>:0
[18:13:57] dbug: 18:13:57.8642640   at Xamarin.Launcher.SimController.InstallAsync (Xamarin.Hosting.SimulatorInstallConfig config) [0x0001c] in <fd345040cf8646afbcdd85e8867162a4>:0
```

**Comment by filipnavara** [comment]
@steveisok That's not too surprising given the code:
```
	IEnumerable<MachOFile> files = MachO.Read(app_bundle.ExecutablePath);
	foreach (MachOFile file in files)
	{
		if (file.Architecture == MachO.Architectures.x86_64)
		{
			return;
		}
	}
```

Unfortunately it's the closed source part of Xamarin so I cannot submit a PR.

**Comment by akoeplinger** [comment]
@mandel-macaque @premun would you mind fixing the mlaunch code to support arm64 simulators?

**Comment by premun** [comment]
@akoeplinger I opened a PR with a following fix:
image

Do you have an ARM64 machine at hand? I can send you the built mlaunch with this fix and you can test the run (using `--mlaunch` arg on XHarness)?

**Comment by akoeplinger** [comment]
Yep I do have an arm64 mac, but I'm about to hop into a meeting. Can you attach the new mlaunch here?

**Comment by premun** [comment]
[mlaunch.tar.gz](https://github.com/dotnet/runtime/files/6344236/mlaunch.tar.gz)
then
```
xharness apple test ... --mlaunch=mlaunch.app/Contents/MacOS/mlaunch
```

File: src/libraries/tests.proj [hunk 1]
```diff
@@ -384,6 +384,9 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="$(RepoRoot)\src\libraries\System.Buffers\tests\System.Buffers.Tests.csproj"
+                      Exclude="@ProjectExclusions" />
+    <!--
```
Comment (lambdageek) [inline src/libraries/tests.proj:389]: This is just temporarily commented out, right?

Comment (steveisok) [reply src/libraries/tests.proj:389]: Yes, once the sanity check passes, I'll enable the rest.

File: src/tasks/AppleAppBuilder/Xcode.cs [hunk 2]
```diff
@@ -87,7 +87,7 @@ public Xcode(string target, string arch)
         var entitlements = new List<KeyValuePair<string, string>>();
 
         bool hardenedRuntime = false;
-        if (Target == TargetNames.MacCatalyst && !(forceInterpreter || forceAOT)) {
+        if (Target == TargetNames.MacCatalyst && !forceAOT) {
             hardenedRuntime = true;
 
             /* for mmmap MAP_JIT */
```
Comment (lambdageek) [inline src/tasks/AppleAppBuilder/Xcode.cs:93]: Can you leave a `/* FIXME: figure out why we are trying to JIT trampolines, disable it, and remove the JIT entitlement when we're using the interpreter */` just before the `entitlements.Add` for the JIT

**Comment by lambdageek** [comment]
We should give this another go using "JustInterp" (`--aot=interp`) mode like in #53197.  The cross-compilers should be in better shape because most of the no-exec stuff was not catalyst-specific.

**Comment by steveisok** [comment]
@lambdageek After a few adjustments on my local build, it's now failing on

```
appdomain.c:197, condition `is_ok (error)' not met, function:create_domain_objects, Attempting to JIT compile method 'void System.OutOfMemoryException:.ctor (string)' while running in aot-only mode.
```

**Comment by lambdageek** [comment]
@steveisok the `src/tasks/AppleAppBuilder/Templates/runtime.m` template needs to be adjusted to set the AOT mode to `MONO_AOT_MODE_INTERP`.

We have all that stuff conditioned on `TARGET_OS_MACCATALYST`, but probably we should have some more direct way to pass down the appropriate AOT mode (instead of trying to guess from the target and from the force aot / force interpreter values) from the Xcode.cs down to the CMakeLists.txt template and down to runtime.m

https://github.com/dotnet/runtime/blob/0416c3469b7bccff087b89db3d9967dfd36070a3/src/tasks/AppleAppBuilder/Templates/runtime.m#L300-L310

**Pull request closed without merging**